### ISSUE TITLE
Vault fails to validate certificate for MongoDB database connection (DocumentDB).

### ISSUE THREAD

**Issue by cyrus-mc** [hashicorp/vault#9509]
**Describe the bug**

When configuring a database secret to mongoDB (in my case AWS DocumentDB) vault errors with "x509: certificate signed by unknown authority" even though I pass in via tls_ca the Certificate Authority used to sign the DocumentD certificates. 

**To Reproduce**

1. Run `vault write database/config/my-mongodb-database plugin_name=mongodb-database-plugin allowed_roles="my_role" connection_url="mongodb://{{username}}:{{password}}@unit-test01.cluster-cz6h536kutdi.xxxxxxx:27017/admin?tls=true" username="master" password="xxxx" tls_ca=@rds-ca-2019-root.pem`

Vault will respond with

```
Error writing data to database/config/my-mongodb-database: Error making API request.

URL: PUT http://localhost:8200/v1/database/config/my-mongodb-database
Code: 400. Errors:

* error creating database object: error verifying connection: server selection error: server selection timeout, current topology: { Type: Unknown, Servers: [{ Addr: unit-test01.cluster-cz6h536kutdi.xxxxxx.com:27017, Type: Unknown, State: Connected, Average RTT: 0, Last error: connection() : x509: certificate signed by unknown authority }, ] }
```

**Expected behavior**

The expectation is that the secret will be successfully added. According to the documentation the value of tls_ca is passed as option to `tlsCAFile` for Mongo. 

```
› mongo --tls --host unit-test01.cluster-cz6h536kutdi.xxxxx.com:27017 -u master -p xxxxx --tlsCAFile rds-ca-2019-root.pem
MongoDB shell version v4.2.8
connecting to: mongodb://unit-test01.cluster-cz6h536kutdi.xxxxxx.com:27017/?compressors=disabled&gssapiServiceName=mongodb
Implicit session: session { "id" : UUID("9058bb07-5d3a-42b6-938a-6ccb4d515a90") }
MongoDB server version: 3.6.0
WARNING: shell and server versions do not match

Warning: Non-Genuine MongoDB Detected

This server or service appears to be an emulation of MongoDB rather than an official MongoDB product.

Some documented MongoDB features may work differently, be entirely missing or incomplete, or have unexpected performance characteristics.

To learn more please visit: https://dochub.mongodb.org/core/non-genuine-mongodb-server-warning.

rs0:PRIMARY>
```

Which as shown above, works if I call mongo shell directly. 

**Environment:**

* Vault Server Version:

```
› vault status
Key             Value
---             -----
Seal Type       shamir
Initialized     true
Sealed          false
Total Shares    1
Threshold       1
Version         1.4.3
Cluster Name    vault-cluster-78863a82
Cluster ID      8da576e2-1f92-c03c-292b-efcbcdd421fb
HA Enabled      false
```

* Vault CLI Version (retrieve with `vault version`):

```
› vault version
Vault v1.4.3
```

* Server Operating System/Architecture:

I have tried this running Vault locally (in dev mode on Mac) as well as within a containerized/Kubernetes setup. Both exhibit the same issue.

**Comment by cyrus-mc** [comment]
I can't figure this out. I figured up the code and debugged and everything looks to be set correctly. Even followed https://docs.aws.amazon.com/documentdb/latest/developerguide/connect_programmatically.html, which is pretty much the same way Vault instantiates a connection.

The example AWS provides connects just fine to the database, but Vault wants to error on the Ping step.

**Comment by cyrus-mc** [comment]
I think I figured out the issue. The issue resolves around the fact that in my connection string I am have `tls=true`. 

Looking through the mongoDB code, what happens is that a call to `ApplyURI` clears the `TLSConfig`

Here is the relevant code

```
func (c *ClientOptions) ApplyURI(uri string) *ClientOptions {
	if c.err != nil {
		return c
	}

...

if cs.SSL {
		tlsConfig := new(tls.Config)

		if cs.SSLCaFileSet {
			c.err = addCACertFromFile(tlsConfig, cs.SSLCaFile)
			if c.err != nil {
				return c
			}
		}

		if cs.SSLInsecure {
			tlsConfig.InsecureSkipVerify = true
		}

		var x509Subject string
		var keyPasswd string
		if cs.SSLClientCertificateKeyPasswordSet && cs.SSLClientCertificateKeyPassword != nil {
			keyPasswd = cs.SSLClientCertificateKeyPassword()
		}
		if cs.SSLClientCertificateKeyFileSet {
			x509Subject, err = addClientCertFromConcatenatedFile(tlsConfig, cs.SSLClientCertificateKeyFile, keyPasswd)
		} else if cs.SSLCertificateFileSet || cs.SSLPrivateKeyFileSet {
			x509Subject, err = addClientCertFromSeparateFiles(tlsConfig, cs.SSLCertificateFile,
				cs.SSLPrivateKeyFile, keyPasswd)
		}
		if err != nil {
			c.err = err
			return c
		}

		// If a username wasn't specified fork x509, add one from the certificate.
		if c.Auth != nil && strings.ToLower(c.Auth.AuthMechanism) == "mongodb-x509" &&
			c.Auth.Username == "" {

			// The Go x509 package gives the subject with the pairs in reverse order that we want.
			c.Auth.Username = extractX509UsernameFromSubject(x509Subject)
		}

		c.TLSConfig = tlsConfig
	}

...
}
```

This ends up clearing the rootCAs that are set in `connection_producer.go` for mongo as the call to ApplyURI is done after the call to MergeClientOptions which results in the TLSConfig being reset and not containing the rootCAs we specified.

**Comment by cyrus-mc** [comment]
The workaround for now is to just not set the option `ssl=true` or `tls=true`. However I think the code should accept a valid  mongoDB connection string so some refactoring of this code would be required.

**Comment by cyrus-mc** [comment]
Fix merged via PR #9519
